File: lib/models/deal.dart
class Deal {
  final String id;
  final String name;
  final String description;
  final double discountPercentage;
  final double? originalPrice;
  final double? discountedPrice;
  final String businessName;
  final String? imageUrl;
  final List<String> tags;
  final DealLocation location;

  Deal({
    required this.id,
    required this.name,
    required this.description,
    required this.discountPercentage,
    this.originalPrice,
    this.discountedPrice,
    required this.businessName,
    this.imageUrl,
    required this.tags,
    required this.location,
  });

  factory Deal.fromJson(Map<String, dynamic> json) {
    return Deal(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      discountPercentage: (json['discountPercentage'] as num).toDouble(),
      originalPrice: (json['originalPrice'] as num?)?.toDouble(),
      discountedPrice: (json['discountedPrice'] as num?)?.toDouble(),
      businessName: json['businessName'],
      imageUrl: json['imageUrl'],
      tags: List<String>.from(json['tags']),
      location: DealLocation.fromJson(json['location']),
    );
  }
}

class DealLocation {
  final String type;
  final List<double> coordinates;

  DealLocation({required this.type, required this.coordinates});

  factory DealLocation.fromJson(Map<String, dynamic> json) {
    return DealLocation(
      type: json['type'],
      coordinates: List<double>.from(json['coordinates'].map((c) => (c as num).toDouble())),
    );
  }

  double get longitude => coordinates[0];
  double get latitude => coordinates[1];
}

File: server/models/deal.js
const mongoose = require('mongoose');

const dealSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
  },
  description: {
    type: String,
    required: true,
  },
  discountPercentage: {
    type: Number,
    required: true,
    min: 0,
    max: 100,
  },
  originalPrice: {
    type: Number,
    required: false,
  },
  businessName: {
    type: String,
    required: true,
  },
  imageUrl: {
    type: String,
    required: false,
  },
  tags: [String],
  location: {
    type: {
      type: String,
      enum: ['Point'],
      required: true
    },
    coordinates: {
      type: [Number],
      required: true
    }
  }
}, {
  timestamps: true,
  toJSON: {
    virtuals: true,
    transform: (doc, ret) => {
      delete ret._id;
      delete ret.__v;
    }
  }
});

dealSchema.index({ location: '2dsphere' });

// Virtual for discountedPrice
dealSchema.virtual('discountedPrice').get(function() {
  if (this.originalPrice) {
    return this.originalPrice * (1 - this.discountPercentage / 100);
  }
  return null;
});

const Deal = mongoose.model('Deal', dealSchema);

module.exports = Deal;
